docs/content/oauth-oidc.md

```diff
@@ -0,0 +1,107 @@
+---
+title: OAuth2 and OpenID Connect
+kind: misc
+weight: 5
+---
+
+OAuth2 and OpenID Connect are both pervasive technologies in modern identity systems. While verification of JSON web tokens issued by these systems is documented in the [policy reference](https://www.openpolicyagent.org/docs/latest/policy-reference/#token-verification), the policy examples below aim to cover some other common use cases.
+
+## Metadata discovery
+
+Rather than storing endpoints and other metadata as part of policy data, the authorization server metadata endpoint may be queried for this data.
+
+```live:oidc:module
+package oidc
+
+issuers = {"https://issuer1.example.com", "https://issuer2.example.com"}
+
+metadata_discovery(issuer) = http.send({
+    "url": concat("", issuers[issuer], "/.well-known/openid-configuration",
+    "method": "GET",
+    "force_cache": true,
+    "force_cache_duration_seconds": 86400 # Cache response for 24 hours
+}).body
+
+claims := jwt.decode(input.token)[1]
+metadata = metadata_discovery(claims.iss)
+
+jwks_endpoint := metadata.jwks_uri
+token_endpoint := metadata.token_endpoint
+```
+
+## Token verification using JWKS endpoint
+
+Below example uses the keys published at the JWKS endpoint of the authorization server for token verification.
+
+```live:oidc2:module
+package oidc
+
+jwks_request(url) = http.send({
+    "url": url,
+    "method": "GET",
+    "force_cache": true,
+    "force_cache_duration_seconds": 3600 # Cache response for an hour
+})
+
+jwks = jwks_request("https://authorization-server.example.com/jwks").body
+
+verified = io.jwt.verify_rs256(input.token, json.marshal(jwks))
+```
+
+### Key rotation
+
+Use the keys published at the JWKS endpoint of the authorization server for token verification, with [key rotation](https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys) taken into account.
+
+```live:oidc3:module
+package oidc
+
+jwks_request(url) = http.send({
+    "url": url,
+    "method": "GET",
+    "force_cache": true,
+    "force_cache_duration_seconds": 3600
+})
+
+jwt_decoded := io.jwt.decode(input.token)
+jwt_header := decoded[0]
+
+jwks = jwks_cached {
+    jwks_cached := jwks_request("https://authorization-server.example.com/jwks").body
+    jwt_header.kid == jwks_cached.keys[_].kid
+} else = jwks_rotated {
+    # Add query param to second request to avoid both getting the same cache key
+    jwks_rotated := jwks_request("https://authorization-server.example.com/jwks?r=1").body
+}
+
+jwt_verified := io.jwt.verify_rs256(input.token, json.marshal(jwks))
```

Comment:
I wonder if something like this would be useful:

```rego
unverified_jwt := io.jwt.decode(input.token)

verified_jwt := unverified_jwt {
  io.jwt.verify_rs256(input.token, json.marshal(jwks))
}
```

Reply:
👍 updated